lib/sshutils/tcpip.go

```diff
@@ -35,7 +42,69 @@ func ParseDirectTCPIPReq(data []byte) (*DirectTCPIPReq, error) {
 	var r DirectTCPIPReq
 	if err := ssh.Unmarshal(data, &r); err != nil {
 		log.Infof("failed to parse Direct TCP IP request: %v", err)
-		return nil, err
+		return nil, trace.Wrap(err)
 	}
 	return &r, nil
 }
+
+type TCPIPForwardRequest struct {
+	Addr string
+	Port uint32
+}
+
+func ParseTCPIPForwardRequest(data []byte) (*TCPIPForwardRequest, error) {
+	var r TCPIPForwardRequest
+	if err := ssh.Unmarshal(data, &r); err != nil {
+		log.Infof("failed to parse TCP IP Forward request: %v", err)
+		return nil, trace.Wrap(err)
+	}
+	return &r, nil
+}
+
+type channelOpener interface {
+	OpenChannel(name string, data []byte) (ssh.Channel, <-chan *ssh.Request, error)
+}
+
+// StartRemoteListener listens on the given listener and forwards any accepted
+// connections over a new "forwarded-tcpip" channel.
+func StartRemoteListener(ctx context.Context, sshConn channelOpener, srcAddr, dstAddr string, listener net.Listener) error {
+	srcHost, srcPort, err := SplitHostPort(srcAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	dstHost, dstPort, err := SplitHostPort(dstAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	req := ForwardedTCPIPRequest{
+		Addr:     srcHost,
+		Port:     srcPort,
+		OrigAddr: dstHost,
+		OrigPort: dstPort,
+	}
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	reqBytes := ssh.Marshal(req)
+
+	go func() {
+		for ctx.Err() == nil {
```

Comment:
The docs for `(*ssh.Client).Listen` say
> The listener must be serviced, or the SSH connection may hang.

What happens if the context is done but the listener isn't?